### ISSUE TITLE
`NoReturn` should be flattened in some fashion

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1183]
**Describe the bug**
Pyright appears to consider `NoReturn` not to be part of the type hierarchy; functions which return `NoReturn` cannot be used in callbacks where the callback is a union of a function returning `NoReturn` and another function returning a concrete type.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That `baz(quux)` would not produce an error.

**Screenshots or Code**

```python
# pyright: strict

from __future__ import annotations

from typing import Callable, NoReturn, TypeVar

T = TypeVar('T')

def foo() -> str:
    ...

def bar() -> NoReturn:
    ...

def baz(value: Callable[..., T]) -> T:
    "Error handling happens here."

async def qux() -> None:
    quux = foo if 1 else bar
    reveal_type(quux)  # Type of "quux" is "() -> str | () -> NoReturn"

    # Argument of type "() -> str | () -> NoReturn" cannot be assigned to parameter "value" of type "(*args: Any, **kwargs: Any) -> T" in function "baz"
    #   Type "() -> str | () -> NoReturn" cannot be assigned to type "(*args: Any, **kwargs: Any) -> T"
    #     Type "() -> NoReturn" cannot be assigned to type "(*args: Any, **kwargs: Any) -> T"
    #       Function return type "NoReturn" is not compatible with type "T"
    #         Type "str" cannot be assigned to type "NoReturn"
    reveal_type(baz(quux))
    reveal_type(baz(bar))

```

**VS Code extension or command-line**
Either, 1.1.86

**Comment by erictraut** [comment]
The problem is not that `NoReturn` is not part of the object hierarchy. It is modeled as a class that derives from `object` (as do all classes). The behavior you're seeing is because of special-case logic that prevents a `NoReturn` from ever appearing in a union. It is filtered out in such cases. You can see this by doing the following:

```python
def func(val: Union[NoReturn, int]):
    reveal_type(val) # int
```

When the TypeVar Solver attempts to solve for `T` in the generic function `baz`, it solves it as `Union[str, NoReturn]` which gets filtered down to just `str`, but then `quux` is no longer assignable to it.

Perhaps this filtering behavior in the general case is not appropriate and NoReturn should be allowed in unions in general but filtered only in cases where return types are being inferred.

What do you think?

**Comment by erictraut** [comment]
Upon further thought, I'm convinced that NoReturn should be allowed in unions except in the limited case where pyright is inferring the return type for a function that doesn't have a declared return type. I've updated the code accordingly, and your sample code above now works as you would expect.

**Comment by layday** [comment]
Is the intention here that `baz(quux)` should return `str | NoReturn`?  I assume one would then have to annotate `baz` as accepting a callable with a return type of `T | NoReturn` to pluck `NoReturn` out of `T`?  That seems reasonable, though it appears to be sensitive to the order of the members in the union as implemented:

```py
# pyright: strict

from __future__ import annotations

from typing import Callable, NoReturn, TypeVar, Union

T = TypeVar('T')

def foo() -> str:
    ...

def bar() -> NoReturn:
    ...

def baz(value: Callable[..., Union[T, NoReturn]]) -> T:
    "Error handling happens here."

def qux() -> None:
    quux = bar if 1 else foo
    reveal_type(quux)
    reveal_type(baz(quux))
    quuux = foo if 1 else bar
    reveal_type(quuux)
    reveal_type(baz(quuux))
```

Output:

```
  20:17 - info: Type of "quux" is "() -> NoReturn | () -> str"
  21:17 - info: Type of "baz(quux)" is "NoReturn | str"
  23:17 - info: Type of "quuux" is "() -> str | () -> NoReturn"
  24:17 - info: Type of "baz(quuux)" is "str"
```

**Comment by erictraut** [comment]
This is now addressed in pyright 1.1.87, which I just published. It will also be included in the next release of pylance.
